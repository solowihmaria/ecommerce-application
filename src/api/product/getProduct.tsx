import axios from 'axios';
import type { ProductResponse } from './product.types';
import type { AuthResponse } from '../createCustomer/createCustomer.types';
import { getToken } from '../token';

export const getUserToken = async (loginStatus: boolean) => {
    let accessToken: string = '';

    if (loginStatus) {
        const storedToken = getToken();
        if (storedToken) {
            accessToken = storedToken;
        }
    } else {
        const tokenResponse = await getGuestToken();
        accessToken = tokenResponse.access_token;
    }

    return accessToken;
};

const getGuestToken = async () => {
    const authUrl = process.env.CTP_AUTH_URL;
    const clientId = process.env.CTP_CLIENT_ID;
    const clientSecret = process.env.CTP_CLIENT_SECRET;

    const tokenUrl = `${authUrl}/oauth/token`;
    const credentials = btoa(`${clientId}:${clientSecret}`);

    const response = await axios.post<AuthResponse>(
        tokenUrl,
        new URLSearchParams({
            grant_type: 'client_credentials',
        }),
        {
            headers: {
                Authorization: `Basic ${credentials}`,
                'Content-type': 'application/x-www-form-urlencoded',
            },
        }
    );

    return response.data;
};

export const getProductByID = async (
    id: string,
    token: string
): Promise<ProductResponse> => {
    const apiUrl = process.env.CTP_API_URL;
    const projectKey = process.env.CTP_PROJECT_KEY;

    const productUrl = `${apiUrl}/${projectKey}/products/${id}`;
    const product = await axios.get<ProductResponse>(productUrl, {
        headers: {
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
    });
    return product.data;
};
