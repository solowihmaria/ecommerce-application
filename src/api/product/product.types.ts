export enum Attributes {
    family = 'family',
    size = 'size',
    careLevel = 'care-level',
    toxic = 'toxic',
    lightRequirements = 'light-requirements',
    height = 'height',
}

export enum Sizes {
    small = 'S',
    medium = 'M',
    large = 'L',
}

export enum Care {
    easy = 'easy',
    medium = 'medium',
    hard = 'hard',
}

export enum Light {
    low = 'low',
    medium = 'medium',
    high = 'high',
}

export interface ProductResponse {
    id: string;
    masterData: {
        current: {
            name: {
                'en-US': string;
            };
            description: {
                'en-US': string;
            };
            slug: {
                'en-US': string;
            };
            masterVariant: Variant;
            variants: Variant[];
        };
    };
}

export interface Variant {
    id: number;
    sku: string;
    prices: Price[];
    images: Image[];
    attributes: (
        | FamilyAttribute
        | SizeAttribute
        | CareAttribute
        | ToxicAttribute
        | LightAttribute
        | HeightAttribute
    )[];
}

export type AttributesList = (
    | [string, string]
    | [string, Sizes]
    | [string, Care]
    | [string, boolean]
    | [string, Light]
    | [string, number]
)[];

export interface Price {
    id: string;
    value: {
        type: string;
        currencyCode: string;
        centAmount: number;
        fractionDigits: number;
    };
    discounted?: {
        value: {
            type: string;
            currencyCode: string;
            centAmount: number;
            fractionDigits: number;
        };
    };
    key: string;
}

export interface Image {
    url: string;
    label: string;
    dimensions: {
        w: number;
        h: number;
    };
}

interface FamilyAttribute {
    name: Attributes.family;
    value: string;
}

interface SizeAttribute {
    name: Attributes.size;
    value: {
        key: Sizes;
        label: Sizes;
    };
}

interface CareAttribute {
    name: Attributes.careLevel;
    value: {
        key: Care;
        label: Care;
    };
}

interface ToxicAttribute {
    name: Attributes.toxic;
    value: boolean;
}

interface HeightAttribute {
    name: Attributes.height;
    value: number;
}

interface LightAttribute {
    name: Attributes.lightRequirements;
    value: {
        key: Light;
        label: Light;
    };
}

export interface CustomProduct {
    id: string;
    name: string;
    description: string;
    masterVariant: CustomVariant;
    variants: CustomVariant[];
}

export interface CustomVariant {
    id: number;
    sku: string;
    price: number;
    images: Image[];
    family: string;
    size: Sizes;
    care: Care;
    toxic: boolean;
    height: number;
    light: Light;
    discount: number | null;
}

export interface CustomAttributes {
    family: string;
    size: Sizes;
    'care-level': Care;
    toxic: boolean;
    height: number;
    'light-requirements': Light;
}
