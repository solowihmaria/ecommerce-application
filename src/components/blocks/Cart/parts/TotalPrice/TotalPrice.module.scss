@use '../../../../../assets/styles/constants.module' as c;

.totalPriceContainer {
    display: flex;
    gap: 1rem;
    align-items: center;
    justify-content: flex-end;
}

.column {
    width: 30%;
}

.columnName {
    color: c.$color-text-primary;
    text-align: center;
}

.columnValue {
    color: c.$color-text-primary;
    text-align: center;
}

.cartTotalOriginContainer {
    position: relative;
    display: inline-flex;
    width: fit-content;
    margin-right: -2rem;
}

.cartTotalOrigin {
    display: inline-block;
    font-size: 24px;
    color: c.$color-gray-600;
}

.cross {
    position: absolute;
    top: 53%;
    display: block;
    width: 95%;
    border-top: 2px solid c.$color-gray-600;
}

.totalPrice {
    display: flex;
    gap: 1rem;
    justify-content: center;
}

.cartTotalNew {
    font-size: 1.5rem;
}

.cartTotalPrice {
    font-size: 1.5rem;
}

.totalLabel {
    font-size: 1.125rem;
}

@media (width <= 999px) {
    .cartTotalOriginContainer {
        margin-right: -1rem;
    }
}

@media (width <= 500px) {
    .cartTotalOriginContainer {
        margin-right: -1rem;
    }
}
