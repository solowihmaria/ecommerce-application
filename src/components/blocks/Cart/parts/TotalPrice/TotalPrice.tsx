import clsx from 'clsx';
import type { CustomCart } from '../../../../../api/cart/cart.types';
import styles from './TotalPrice.module.scss';
import { getCartTotalOrigin } from '../../helpers/totalPriceHelper';

export const TotalPrice = ({ cartContent }: { cartContent: CustomCart }) => {
    const isAnyProductOnSale = (cartContent: CustomCart) => {
        return cartContent.lineItems.some((item) => item.variant.discount);
    };
    return (
        <div
            data-testid="total-attribute"
            className={styles.totalPriceContainer}
        >
            <p
                className={clsx(
                    styles.columnValue,
                    styles.column,
                    styles.totalLabel
                )}
            >
                Total(â‚¬):
            </p>
            {isAnyProductOnSale(cartContent) ||
            cartContent.discountOnTotalPrice ||
            cartContent.lineItems[0].discountedPricePerQuantity ? (
                <>
                    <p
                        data-testid="total-attribute-with-sale-old"
                        className={clsx(
                            styles.columnValue,
                            styles.column,
                            styles.cartTotalOriginContainer
                        )}
                    >
                        <span className={styles.cartTotalOrigin}>
                            {getCartTotalOrigin(cartContent)}
                        </span>
                        <span className={clsx(styles.cross)}></span>
                    </p>

                    <p
                        data-testid="total-attribute-with-sale-new"
                        className={clsx(
                            styles.columnValue,
                            styles.column,
                            styles.cartTotalNew
                        )}
                    >
                        {cartContent.totalPrice.toFixed(2)}
                    </p>
                </>
            ) : (
                <p
                    data-testid="total-attribute-without-sale"
                    className={clsx(
                        styles.columnValue,
                        styles.column,
                        styles.cartTotalPrice
                    )}
                >
                    {cartContent.totalPrice.toFixed(2)}
                </p>
            )}
        </div>
    );
};
