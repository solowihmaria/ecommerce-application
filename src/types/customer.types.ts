export enum CountryCode {
    FR = 'FR',
    DE = 'DE',
    IT = 'IT',
}

export type AddressType = 'shipping' | 'billing';

export interface Address {
    id?: string;
    key?: string;
    streetName: string;
    postalCode: string;
    city: string;
    country: CountryCode;
    type?: AddressType;
    isDefault?: boolean;
}

export interface BaseCustomer {
    email: string;
    firstName: string;
    lastName: string;
    dateOfBirth?: string; // Опционально в профиле, но обязательно при регистрации
    addresses: Address[];
    defaultShippingAddressId?: string | null;
    defaultBillingAddressId?: string | null;
}
