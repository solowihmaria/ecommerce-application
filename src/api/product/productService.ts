import { getProductByID } from './getProduct';
import type {
    AttributesList,
    Care,
    CustomAttributes,
    CustomProduct,
    CustomVariant,
    Light,
    ProductResponse,
    Sizes,
    Variant,
} from './product.types';

export const getProductData = async (id: string) => {
    const productData = await getProductByID(id);

    return productData;
};

const prepareVariant = (initialVariant: Variant): CustomVariant => {
    const { id, sku, images, prices } = initialVariant;
    //filter out not used attributes
    const attributes = initialVariant.attributes.filter((attribute) =>
        [
            'family',
            'size',
            'care-level',
            'toxic',
            'light-requirements',
            'height',
        ].includes(attribute.name)
    );
    //get pairs of key value ["attribute name", value]
    const transformedAttributes: AttributesList = attributes.map(
        (attribute) => {
            let value: string | number | boolean | Sizes | Care | Light;
            if (
                typeof attribute.value === 'object' &&
                'label' in attribute.value
            ) {
                value = attribute.value.label;
            } else {
                value = attribute.value;
            }
            return [attribute.name, value];
        }
    );
    const attributesObject: CustomAttributes = Object.fromEntries(
        transformedAttributes
    );

    const discount = prices[0].discounted
        ? prices[0].discounted.value.centAmount / 100
        : null;
    return {
        id: id,
        sku: sku,
        price: prices[0].value.centAmount / 100,
        discount: discount,
        images: images,
        family: attributesObject.family,
        size: attributesObject.size,
        care: attributesObject['care-level'],
        toxic: attributesObject.toxic,
        light: attributesObject['light-requirements'],
        height: attributesObject.height,
    };
};

export const transformProductData = (productData: ProductResponse) => {
    const {
        id,
        masterData: {
            current: { masterVariant, variants, name, description },
        },
    } = productData;

    const product: CustomProduct = {
        id: id,
        name: name['en-US'],
        description: description['en-US'],
        masterVariant: prepareVariant(masterVariant),
        variants: variants.map((variant) => prepareVariant(variant)),
    };

    return product;
};
