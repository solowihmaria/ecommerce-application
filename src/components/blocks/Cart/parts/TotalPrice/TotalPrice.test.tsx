import { render, screen } from '@testing-library/react';

import '@testing-library/jest-dom';
import type { CustomCart } from '../../../../../api/cart/cart.types';
import { Care, Light, Sizes } from '../../../../../types/product.types';
import { TotalPrice } from './TotalPrice';

const cartDataWithoutSale: CustomCart = {
    id: 'fc0e74c0-4c34-4fe0-9e93-192960bffc70',
    customerId: 'cd5a7b69-fa67-4ec6-bcc8-a65a846ab202',
    version: 6,
    lineItems: [
        {
            id: '594a55d1-ce76-46f7-ab10-1d02757e7a06',
            productId: '1ac83f8e-06b3-4ce7-b6c9-c6eb8af28f78',
            name: 'Polka Dot Begonia',
            quantity: 2,
            totalPrice: 44,
            variant: {
                id: 1,
                sku: 'BEGONIA032S',
                price: 22,
                discount: null,
                images: [
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-rtB-G5WT.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-2-nUiiWwzr.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-4-B7ePuLdA.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-3--szdI4it.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                ],
                family: 'Begonia (Maculata)',
                size: Sizes.small,
                care: Care.medium,
                toxic: true,
                light: Light.low,
                height: 30,
            },
            discountedPricePerQuantity: null,
        },
        {
            id: '6c0eb8ef-0913-4918-9688-eb841c1a34eb',
            productId: 'd711295e-8b08-48a9-ad71-d72bfc709198',
            name: 'Nerve Plant',
            quantity: 1,
            totalPrice: 12,
            variant: {
                id: 1,
                sku: 'FITTONIA033S',
                price: 12,
                discount: null,
                images: [
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-raO6u9Q8.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-2-rAd-Mb2a.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-3-VRRl5yGn.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                ],
                family: 'Fittonia (Albivenis)',
                size: Sizes.small,
                care: Care.medium,
                toxic: true,
                light: Light.low,
                height: 10,
            },
            discountedPricePerQuantity: null,
        },
    ],
    totalLineItemQuantity: 3,
    totalPrice: 56,
    discountCodes: [],
    discountOnTotalPrice: null,
};

const cartDataWithSale: CustomCart = {
    id: 'fc0e74c0-4c34-4fe0-9e93-192960bffc70',
    customerId: 'cd5a7b69-fa67-4ec6-bcc8-a65a846ab202',
    version: 17,
    lineItems: [
        {
            id: '594a55d1-ce76-46f7-ab10-1d02757e7a06',
            productId: '1ac83f8e-06b3-4ce7-b6c9-c6eb8af28f78',
            name: 'Polka Dot Begonia',
            quantity: 2,
            totalPrice: 44,
            variant: {
                id: 1,
                sku: 'BEGONIA032S',
                price: 22,
                discount: null,
                images: [
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-rtB-G5WT.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-2-nUiiWwzr.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-4-B7ePuLdA.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Polka%20Dot%20Begonia-3--szdI4it.jpg',
                        label: 'Polka Dot Begonia',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                ],
                family: 'Begonia (Maculata)',
                size: Sizes.small,
                care: Care.medium,
                toxic: true,
                light: Light.high,
                height: 30,
            },
            discountedPricePerQuantity: null,
        },
        {
            id: '6c0eb8ef-0913-4918-9688-eb841c1a34eb',
            productId: 'd711295e-8b08-48a9-ad71-d72bfc709198',
            name: 'Nerve Plant',
            quantity: 1,
            totalPrice: 12,
            variant: {
                id: 1,
                sku: 'FITTONIA033S',
                price: 12,
                discount: null,
                images: [
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-raO6u9Q8.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-2-rAd-Mb2a.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Nerve%20Plant-3-VRRl5yGn.jpg',
                        label: 'Nerve Plant',
                        dimensions: {
                            w: 1080,
                            h: 1080,
                        },
                    },
                ],
                family: 'Fittonia (Albivenis)',
                size: Sizes.small,
                care: Care.medium,
                toxic: true,
                light: Light.high,
                height: 10,
            },
            discountedPricePerQuantity: null,
        },
        {
            id: 'ceb1222a-7c64-40a5-a9cb-0211879fe6cf',
            productId: 'f6f3900d-d29f-4e8f-a84e-af150b3a07a2',
            name: 'Fiddle Leaf Fig',
            quantity: 3,
            totalPrice: 94.5,
            variant: {
                id: 1,
                sku: 'FIDDLE002S',
                price: 35,
                discount: 31.5,
                images: [
                    {
                        url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/FiddleleafFigBush-BHhgm44f.png',
                        label: 'Fiddle Leaf Fig',
                        dimensions: {
                            w: 1000,
                            h: 1000,
                        },
                    },
                ],
                family: 'Ficus (Lyrata)',
                size: Sizes.small,
                care: Care.medium,
                toxic: true,
                light: Light.high,
                height: 45,
            },
            discountedPricePerQuantity: null,
        },
    ],
    totalLineItemQuantity: 6,
    totalPrice: 150.5,
    discountCodes: [],
    discountOnTotalPrice: null,
};

describe('<TotalPrice/> component', () => {
    test('Render TotalPrice component when no products on sale', () => {
        render(<TotalPrice cartContent={cartDataWithoutSale} />);

        const totalContainer = screen.getByTestId('total-attribute');
        const totalElement = screen.getByTestId('total-attribute-without-sale');

        expect(totalContainer).toBeInTheDocument();
        expect(totalElement).toBeInTheDocument();
        expect(totalContainer).toMatchSnapshot();
    });

    test('Render TotalPrice component when there are products on sale', () => {
        render(<TotalPrice cartContent={cartDataWithSale} />);

        const totalContainer = screen.getByTestId('total-attribute');
        const totalElementOld = screen.getByTestId(
            'total-attribute-with-sale-old'
        );
        const totalElementNew = screen.getByTestId(
            'total-attribute-with-sale-new'
        );

        expect(totalContainer).toBeInTheDocument();
        expect(totalElementOld).toBeInTheDocument();
        expect(totalElementNew).toBeInTheDocument();

        expect(totalContainer).toMatchSnapshot();
    });
});
