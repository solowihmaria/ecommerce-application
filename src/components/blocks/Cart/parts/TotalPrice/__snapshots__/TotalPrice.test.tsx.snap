// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TotalPrice/> component Render TotalPrice component when no products on sale 1`] = `
<div
  class="totalPriceContainer"
  data-testid="total-attribute"
>
  <p
    class="columnValue column totalLabel"
  >
    Total(€):
  </p>
  <p
    class="columnValue column cartTotalPrice"
    data-testid="total-attribute-without-sale"
  >
    56.00
  </p>
</div>
`;

exports[`<TotalPrice/> component Render TotalPrice component when there are products on sale 1`] = `
<div
  class="totalPriceContainer"
  data-testid="total-attribute"
>
  <p
    class="columnValue column totalLabel"
  >
    Total(€):
  </p>
  <p
    class="columnValue column cartTotalOriginContainer"
    data-testid="total-attribute-with-sale-old"
  >
    <span
      class="cartTotalOrigin"
    >
      161.00
    </span>
    <span
      class="cross"
    />
  </p>
  <p
    class="columnValue column cartTotalNew"
    data-testid="total-attribute-with-sale-new"
  >
    150.50
  </p>
</div>
`;
